### PR TITLE
Split unit_for into a bang and non-bang version

### PR DESCRIPTION
**Pull request by thegedge**
From @nicolaslupien's [comment](https://github.com/Shopify/measured/pull/51#discussion_r92469298), we should suffix this function with an exclamation mark, since it raises. To keep things super tight, I've also went ahead with his suggestion to split the function. Unit tests already exist to check if `to_unit_name` and `convert` raise.

### CODE REVIEW

File: lib/measured/conversion.rb [hunk 1]
```diff
@@ -48,6 +48,9 @@ def conversion_table
 
   def unit_for(name)
     unit = @units.find { |unit| unit.names_include?(name, case_sensitive: @case_sensitive) }
```
Comment (kmcphillips) [inline lib/measured/conversion.rb]: Assignment is unnecessary here then.

File: lib/measured/conversion.rb [hunk 2]
```diff
@@ -48,6 +48,9 @@ def conversion_table
...
-    unit or raise Measured::UnitError, "Cannot find unit for #{name}"
+  end
+
+  def unit_for!(name)
+    unit_for(name) or raise Measured::UnitError, "Cannot find unit for #{name}"
```
Comment (kmcphillips) [inline lib/measured/conversion.rb]: Avoid `or` entirely. It reads nicely, but the subtle order of operations difference is confusing and causes nasty bugs (ie. lower precedence than `=` for example). It's pretty [well accepted ruby style](https://github.com/bbatsov/ruby-style-guide#no-and-or-or) to just not use them.

Comment (thegedge) [inline lib/measured/conversion.rb]: I'm okay with not using `or`, but a couple of my own thoughts that I wouldn't mind some thoughts on before not using it:

1. The style guide takes a dogmatic approach, banning it without reason. I am vehemently against any style guide statement that lacks reasoning (unless it's obvious, which in this case it isn't). 
2. I've seen style guides that allow it for control flow but not as a boolean operation, which makes a lot more sense. I actually very much disagree with the style guide suggestion of using `||` for control flow, because the precedence of `and` and `or` is very close to that of modifier `if` and modifier `unless`.

tl;dr I don't (completely) agree with the style guide, and in some sense would argue that it's contradicting itself by recommending modifier if/unless but blindly banning or/and.

Comment (nicolaslupien) [inline lib/measured/conversion.rb]: The benefit of choosing a single operator is more pragmatic. It looks obvious when you write it but not necessarily when you review it. Make it clear and don't make me think, even if it needs a few more lines of code.

Comment (thegedge) [inline lib/measured/conversion.rb]: Fair enough. Again, I'm not in total agreement, but it's probably a good idea for us to stay in line with community style guides right now and change it upstream.

Just to be clear about why I'm in disagreement:

1. We're conflating boolean operators with control flow structures (unfortunately I think this is a symptom of many high-level programming languages).
2. The example from the style guide is actually the less intuitive structure (it should be `document.save! unless document.saved?`).
3. It seems weird, almost contradictory, to allow writing `do_stuff if foo = bar(baz)` but not `foo = bar(baz) or raise BarFailedError`.

**Comment by thegedge** [comment]
Updated!
